Perform HTTP fallback in UIProcess and after redirect

sysrqb opened the pull request:
#### cbed04cca06ff94b8b5fb316c54df7e143969ce0
<pre>
Perform HTTP fallback in UIProcess and after redirect
<a href="https://bugs.webkit.org/show_bug.cgi?id=279048">https://bugs.webkit.org/show_bug.cgi?id=279048</a>
<a href="https://rdar.apple.com/133794636">rdar://133794636</a>

Reviewed by Alex Christensen.

HTTP fallback is currently handled in FrameLoader, but that causes problems
with PSON (and eventually will cause problems with site isolation). In
particular, there are currently bad interactions with cross-site redirects and
HTTP fallback. This change moves the controlling logic out of FrameLoader
into WebPageProxy::didiFailProvisionLoad where the load is re-initiated.

This change also allows falling back to HTTP after a redirect. Preventing
fallback after a redirect was overly restrictive and caused some site breakage.
Instead, we restrict this fallback to requests that were optimistically
upgraded.

Finally, WKNavigation.HTTPSFirstWithHTTPRedirect is adding coverage with
HTTPSFirst for how we behave when we fallback to HTTP after upgrading and
following multiple redirects. This is similar to an existing test we already
have for HTTPSOnly.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
(WebCore::FrameLoader::handleLoadFailureRecovery): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, HTTPSFirstHTTPDowngradeAfterPSON)):
(TEST(WKNavigation, HTTPSFirstWithHTTPRedirect)):

Canonical link: <a href="https://commits.webkit.org/283158@main">https://commits.webkit.org/283158@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/03469f8957314995cc98c05c86aca2481af4bdc6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65342 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44711 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69366 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15948 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67460 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52466 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11026 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68408 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41305 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56552 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33091 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37978 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13928 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14825 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59834 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14267 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71071 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9294 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13722 "Found 1 new test failure: imported/w3c/web-platform-tests/media-source/mediasource-activesourcebuffers.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59792 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9326 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56617 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60067 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7674 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1331 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9918 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40521 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41598 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42779 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41342 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6497,6 +6497,13 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
             callClientFunctions();
+    } else {
+        if (auto* websitePolicies = navigation ? navigation->websitePolicies() : nullptr; websitePolicies
+            && (websitePolicies->advancedPrivacyProtections().contains(AdvancedPrivacyProtections::HTTPSFirst) || m_preferences->httpSByDefaultEnabled())
```
Should the capitalization of httpSByDefaultEnabled be httpsByDefaultEnabled?

sysrqb replied:
Yes, but I'll need to adjust the generator script, and I'm going to remove this pref soon, so I decided against making that change.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6497,6 +6497,13 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
 #endif
             callClientFunctions();
+    } else {
+        if (auto* websitePolicies = navigation ? navigation->websitePolicies() : nullptr; websitePolicies
```
Can this be a RefPtr instead of auto*?

sysrqb replied:
Looks like yes.

achristensen07 commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -6452,6 +6452,8 @@ void WebPageProxy::didFailProvisionalLoadForFrameShared(Ref<WebProcessProxy>&& p
     ASSERT(!m_failingProvisionalLoadURL);
     m_failingProvisionalLoadURL = provisionalURL;
+    URL failedURL { provisionalURL };
+    bool canFallbackToHTTP = frame.isMainFrame() && error.errorRecoveryMethod() == ResourceError::ErrorRecoveryMethod::HTTPFallback && failedURL.protocolIs("https"_s);
```
Can we move the scope of this into the WebsitePolicies if statement below and add a nested if statement?  That'll make the code do less in the common case and be better organized so it'll be more clear what these things are used for.

sysrqb replied:
Sure. I moved it here so it is shared between both blocks below (HTTPSOnly and HTTPSFirst), but I can also just duplicate this within each block.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -2280,6 +2280,77 @@ HTTPServer httpServer({
+
+    EXPECT_NULL([webView _safeBrowsingWarning]);
+    TestWebKitAPI::Util::run(&finishedSuccessfully);
+    EXPECT_NULL([webView _safeBrowsingWarning]);
```
Can we make SPI something like _warningView?  This is no longer a safe browsing warning.  Also, could you remove the last forward declaration of _WKSafeBrowsingWarning?

sysrqb replied:
I'll make this change in another PR

Commit:
Perform HTTP fallback in UIProcess and after redirect
https://bugs.webkit.org/show_bug.cgi?id=279048
rdar://133794636

Reviewed by Alex Christensen.

HTTP fallback is currently handled in FrameLoader, but that causes problems
with PSON (and eventually will cause problems with site isolation). In
particular, there are currently bad interactions with cross-site redirects and
HTTP fallback. This change moves the controlling logic out of FrameLoader
into WebPageProxy::didiFailProvisionLoad where the load is re-initiated.

This change also allows falling back to HTTP after a redirect. Preventing
fallback after a redirect was overly restrictive and caused some site breakage.
Instead, we restrict this fallback to requests that were optimistically
upgraded.

Finally, WKNavigation.HTTPSFirstWithHTTPRedirect is adding coverage with
HTTPSFirst for how we behave when we fallback to HTTP after upgrading and
following multiple redirects. This is similar to an existing test we already
have for HTTPSOnly.

* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::checkLoadCompleteForThisFrame):
(WebCore::FrameLoader::handleLoadFailureRecovery): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::didFailProvisionalLoadForFrameShared):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST(WKNavigation, HTTPSFirstHTTPDowngradeAfterPSON)):
(TEST(WKNavigation, HTTPSFirstWithHTTPRedirect)):

Canonical link: https://commits.webkit.org/283158@main
